bundler/spec/bundler/ui/json_shell_spec.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+RSpec.describe Bundler::UI::JsonShell do
+  subject { described_class.new }
+
+  before { subject.level = "debug" }
+
+  describe "#info" do
+    before { subject.level = "info" }
+    it "prints to stderr" do
+      expect { subject.info("info") }.to output("info\n").to_stderr
+    end
+  end
+
+  describe "#confirm" do
+    before { subject.level = "confirm" }
+    it "prints to stderr" do
+      expect { subject.confirm("confirm") }.to output("confirm\n").to_stderr
+    end
+  end
+
+  describe "#warn" do
+    before { subject.level = "warn" }
+    it "prints to stderr" do
+      expect { subject.warn("warning") }.to output("warning\n").to_stderr
+    end
+  end
+
+  describe "#debug" do
+    it "prints to stderr" do
+      expect { subject.debug("debug") }.to output("debug\n").to_stderr
+    end
+  end
+
+  describe "#error" do
+    before { subject.level = "error" }
+
+    it "prints to stderr" do
+      expect { subject.error("error!!!") }.to output("error!!!\n").to_stderr
+    end
+
+    context "when stderr is closed" do
+      it "doesn't report anything" do
+        output = begin
+                   result = StringIO.new
+                   result.close
+
+                   $stderr = result
+
+                   subject.error("Something went wrong")
+
+                   result.string
+                 ensure
+                   $stderr = STDERR
+                 end
+        expect(output).to_not eq("Something went wrong")
+      end
+    end
+  end
+
+  describe "#table" do
+    before { subject.level = "info" }
+    let(:header) { { :foo => "Foo", :bar => "Bar (but longer)", :baz => "Baz" } }
+    let(:records) do
+      [
+        { :foo => "Entry 1", :bar => 5, :baz => "a few words" },
+        { :foo => :two, :bar => "five", :baz => "less", :huh => "extra data" },
+        { :foo => 3, :baz => "not as long" },
+      ]
+    end
+
+    it "constructs and prints the table to stdout properly" do
+      # * includes a nil-values entry for the missing :bar
+      # * includes the non-header-ed :huh key
+      expanded_json =
+        <<~JSON
+          [
+          {"foo":"Entry 1","bar":5,"baz":"a few words"},
+          {"foo":"two","bar":"five","baz":"less","huh":"extra data"},
```

Comment:
I like that it prints extra keys for extended data compared to `table`. Can we change `huh` to something like `"extra_keys" => "display extended data beyond table"`. I initially skipped the comment so I thought it was a mistake that `huh` was expected.

Reply:
That's definitely clearer (and yeah, now that it'll support them I'll add that as a case)